packages/std/src/results/response.rs

```diff
@@ -108,36 +101,69 @@ where
     }
 
     /// Add an attribute included in the main `wasm` event.
-    pub fn add_attribute<K: Into<String>, V: Into<String>>(&mut self, key: K, value: V) {
-        self.attributes.push(Attribute {
-            key: key.into(),
-            value: value.into(),
-        });
+    pub fn add_attribute(mut self, attr: impl Into<Attribute>) -> Self {
+        self.attributes.push(attr.into());
+        self
     }
 
     /// This creates a "fire and forget" message, by using `SubMsg::new()` to wrap it,
     /// and adds it to the list of messages to process.
-    pub fn add_message<U: Into<CosmosMsg<T>>>(&mut self, msg: U) {
+    pub fn add_message(mut self, msg: impl Into<CosmosMsg<T>>) -> Self {
         self.messages.push(SubMsg::new(msg));
+        self
     }
 
     /// This takes an explicit SubMsg (creates via eg. `reply_on_error`)
     /// and adds it to the list of messages to process.
-    pub fn add_submessage(&mut self, msg: SubMsg<T>) {
+    pub fn add_submessage(mut self, msg: SubMsg<T>) -> Self {
         self.messages.push(msg);
+        self
     }
 
     /// Adds an extra event to the response, separate from the main `wasm` event
     /// that is always created.
     ///
     /// The `wasm-` prefix will be appended by the runtime to the provided type
     /// of event.
-    pub fn add_event(&mut self, event: Event) {
+    pub fn add_event(mut self, event: Event) -> Self {
         self.events.push(event);
+        self
+    }
+
+    /// Bulk add attributes included in the main `wasm` event.
+    pub fn add_attributes<A: Into<Attribute>>(
```

Comment:
Ah, you did add this.

And this is where you really want to make use of the `Into<Attribute>` to accept `Vec<Attribute>`, `Vec<(&str, &str)>`, as well as `Vec<(&str, String)>`. This makes sense, but not so much with `add_attribute`.

Is this used in some test contracts? I have seen multiple `add_attribute` calls chained together

Reply:
I cannot find one place where it can be used in the example contracts.
The idea is good. Even better if each item in the iterator could be a different type, but all implement `Into<Attribute>`. I guess this might be possible with `Box<dyn Into<Attribute>>` but make usage even more annoying.

In any case, it would be good to have a test or two to document how this could be used (maybe 3 attributes of the same type, 1 of different added with the other function)?

Reply:
I can do that. I found a use for `add_submessages`, created that method, and then thought "might as well do the same for all the other collection fields for consistency." If you think some of these are unnecessary, I can get rid of them.

I feel like yeah, trait objects would likely only make this overcomplicated. Maybe if someone asks for it one day we'll look into it?

Reply:
Comments above to add an example making this from `Vec<Attribute>`, showing some way of doing so.